### PR TITLE
Refactor Players Controller

### LINKED ISSUE #169: Add `/players` `PATCH` endpoint
Allow administrators to modify `Player` fields.

Namely, this is specifically useful for applying some kind of restriction status as proposed in #130.

### LINKED ISSUE #196: Refactor `/api/v1/matches/player/{osuId}` into `/api/v1/players/{key}/matches`
`key` here is a dynamic search which we already have an implementation for. Aids #138

### PR DESCRIPTION
**Pull request by myssto**
Another one bites the dust

## Breaking Changes
- **Change** `PlayerRanksDTO` (Returns from `/players/ranks`)
  - **Add** field `string? username`
  - **Add** field `string? country`
  - **Rename** field `earliestOsuGlobalRank` to `earliestStandardGlobalRank`
  - **Rename** field `earliestOsuGlobalRankDate` to `earliestStandardGlobalRankDate`
- **Remove** `GET` `/players/id-mapping`
- **Remove** `GET` `/players/country-mapping`
- **Move** `GET` `/matches/player/{osuId}` to `/players/{key}/matches`
- **Rename** `GET` `/players/all` to `/players`
- **Rename** `GET` `/players/ranks/all` to `/players/ranks`
- **Rename** `GET` `/players/{key}/info` to `/players/{key}`

## Additions
- **Add** `PATCH` `/players/{id}`

Completes task `Refactor PlayersController` of #207
Closes #169
Closes #196

### CODE REVIEW

**Comment by myssto** [comment]
This is essentially a *complete rewrite* of the players route. If you depend on any of the endpoints, please be aware. I will open issues on processor and web soon. @AkinariHex @486c @S1n1st3rn3ss

**Comment by hburn7** [comment]
> This is essentially a _complete rewrite_ of the players route. If you depend on any of the endpoints, please be aware. I will open issues on processor and web soon. @AkinariHex @486c @S1n1st3rn3ss

I know for a fact all of the removed routes are used by the processor. What is the plan for these?

File: API/DTOs/PlayerRanksDTO.cs [hunk 1]
```diff
@@ -1,20 +1,67 @@
...
+
+    /// <summary>
+    /// Current mania rank for the player
+    /// </summary>
     public int? RankMania { get; set; }
 
-    public int? EarliestOsuGlobalRank { get; set; }
-    public DateTime? EarliestOsuGlobalRankDate { get; set; }
+    /// <summary>
+    /// Earliest highest standard rank available for the player
```
Comment (hburn7) [inline API/DTOs/PlayerRanksDTO.cs:29]:
```suggestion
    /// Earliest known standard rank available for the player after they started playing tournaments
```

Use similar language throughout

File: API/Repositories/Interfaces/IPlayerRepository.cs [hunk 2]
```diff
@@ -1,88 +1,64 @@
...
+    Task<Player?> GetVersatileAsync(string key);
 
     /// <summary>
-    /// Returns all players
+    /// Gets a list of players for the given username
     /// </summary>
-    /// <param name="eagerLoad">Whether to also load related fields (i.e. player matches)</param>
-    /// <returns></returns>
-    Task<IEnumerable<Player>> GetAsync(bool eagerLoad = false);
+    /// <remarks>Username is case insensitive</remarks>
```
Comment (hburn7) [inline API/Repositories/Interfaces/IPlayerRepository.cs:23]:
```suggestion
    /// <remarks>Username is case insensitive and partially matched</remarks>
```

**Comment by myssto** [comment]
> I know for a fact `/players/id-mapping` and `/players/country-mapping` are used by the processor. What is the plan for these?

I thought I was told that they weren't... maybe we can discuss in dev? I know processor guys suggested better options or said that they were redundant in the first place

**Review by hburn7 — CHANGES_REQUESTED** [review]

**Comment by hburn7** [comment]
> > I know for a fact `/players/id-mapping` and `/players/country-mapping` are used by the processor. What is the plan for these?
> 
> I thought I was told that they weren't... maybe we can discuss in dev? I know processor guys suggested better options or said that they were redundant in the first place

Hm, okay. Still it'd probably be fine to just fetch all player data and make hashmaps from that instead of using these mapping endpoints.

**Comment by myssto** [comment]
Updated breaking changes section to better reflect changes to `PlayerRanksDTO`

File: API/Repositories/Interfaces/IPlayerRepository.cs [hunk 3]
```diff
@@ -42,80 +44,52 @@ public interface IPlayerRepository : IRepository<Player>
...
+    /// <returns>A country code, or null if not found</returns>
+    Task<string?> GetCountryAsync(int id);
 
     /// <summary>
-    /// Returns the player id for the given user id
+    /// Gets all players that haven't been updated in the last 14 days
+    /// or have never been updated
     /// </summary>
-    /// <param name="userId"></param>
-    Task<int> GetIdAsync(int userId);
```
Comment (hburn7) [inline API/Repositories/Interfaces/IPlayerRepository.cs:114]: This introduces a bug. This causes the service method to use the `GetAsync(long osuId)` repository method instead of the correct `int id` before. Why was this removed?
```cs
// PlayerService.cs L:23-24 
public async Task<PlayerDTO?> GetAsync(int id) =>
        _mapper.Map<PlayerDTO?>(await _playerRepository.GetAsync(id)); // !!calling long osuId overload!!
```

This causes the `playerInfo` field in `PlayerStatsService` to be null, breaking user profiles.

**Review by hburn7 — CHANGES_REQUESTED** [review]

**Comment by myssto** [comment]
Closing this for being way too big in scope. Will use as a reference for a piece-wise refactor of the players controller.
